src/SOS/Strike/clrma/exception.cpp

```diff
@@ -0,0 +1,551 @@
+    m_lRefs(1),
+    m_managedAnalysis(managedAnalysis),
+    m_address(address),
+    m_typeName(nullptr),
+    m_message(nullptr),
+    m_exceptionDataInitialized(false),
+    m_stackFramesInitialized(false),
+    m_innerExceptionsInitialized(false)
+{
+    _ASSERTE(address != 0);
+    managedAnalysis->AddRef();
+}
+
+ClrmaException::~ClrmaException()
+{
+    TraceInformation("~ClrmaException\n");
+    if (m_typeName != nullptr)
+    {
+        delete [] m_typeName;
+        m_typeName = nullptr;
+    }
+    if (m_message != nullptr)
+    {
+        delete [] m_message;
+        m_message = nullptr;
+    }
+    if (m_managedAnalysis != nullptr)
+    {
+        m_managedAnalysis->Release();
+        m_managedAnalysis = nullptr;
+    }
+}
+
+/// <summary>
+/// This called by each clrma exception methods to initialize and cache the exception data.
+/// </summary>
+HRESULT
+ClrmaException::Initialize()
+{
+    if (m_managedAnalysis == nullptr)
+    {
+        return E_UNEXPECTED;
+    }
+    if (!m_exceptionDataInitialized)
+    {
+        TraceInformation("ClrmaException::Initialize %016llx\n", m_address);
+
+        HRESULT hr;
+        DacpObjectData objData;
+        if (FAILED(hr = objData.Request(m_managedAnalysis->SosDacInterface(), m_address)))
+        {
+            TraceError("ClrmaException::Initialize GetObjectData FAILED %08x\n", hr);
+            return hr;
+        }
+
+        if (m_managedAnalysis->IsExceptionObj(objData.MethodTable) != 0)
+        {
+            if (FAILED(hr = m_exceptionData.Request(m_managedAnalysis->SosDacInterface(), m_address)))
+            {
+                TraceError("ClrmaException::Initialize GetObjectExceptionData FAILED %08x\n", hr);
+                return hr;
+            }
+            UINT cbTypeName = 0;
+            if (SUCCEEDED(hr = m_managedAnalysis->SosDacInterface()->GetMethodTableName(objData.MethodTable, 0, nullptr, &cbTypeName)))
+            {
+                ArrayHolder<WCHAR> typeName = new (std::nothrow)WCHAR[cbTypeName];
+                if (typeName != nullptr)
+                {
+                    if (SUCCEEDED(hr = m_managedAnalysis->SosDacInterface()->GetMethodTableName(objData.MethodTable, cbTypeName, typeName, nullptr)))
+                    {
+                        m_typeName = typeName.Detach();
+                    }
+                    else
+                    {
+                        TraceError("ClrmaException::Initialize GetMethodTableName(%016x) 2 FAILED %08x\n", objData.MethodTable, hr);
+                    }
+                }
+            }
+            else
+            {
+                TraceError("ClrmaException::Initialize GetMethodTableName(%016x) 1 FAILED %08x\n", objData.MethodTable, hr);
+            }
+            if (m_exceptionData.Message == 0)
+            {
+                // To match the built-in SOS provider that scrapes !pe output.
+                m_message = new (std::nothrow)WCHAR[8];
+                wcscpy(m_message, L"<none>");
```

Comment:
nit: use const wchar_t* for L"<none>", then use `countof` for string allocation - this prevents off-by one errors